File: go/pkce_test.go
package pinto

import (
	"testing"
)

func TestPKCEUtils(t *testing.T) {
	t.Run("GenerateRandomString length", func(t *testing.T) {
		s1 := GenerateRandomString(16)
		s2 := GenerateRandomString(32)
		if len(s1) == 0 || len(s2) == 0 {
			t.Fatal("generated string should not be empty")
		}
		if s1 == s2 {
			t.Fatal("random strings should be distinct")
		}
	})

	t.Run("GenerateCodeVerifier clamp", func(t *testing.T) {
		short := GenerateCodeVerifier(10)
		normal := GenerateCodeVerifier(64)
		long := GenerateCodeVerifier(200)

		if len(short) != 43 {
			t.Fatalf("expected 43, got %d", len(short))
		}
		if len(normal) != 64 {
			t.Fatalf("expected 64, got %d", len(normal))
		}
		if len(long) != 128 {
			t.Fatalf("expected 128, got %d", len(long))
		}
	})

	t.Run("ComputeCodeChallenge against RFC 7636 Appendix B", func(t *testing.T) {
		verifier := "dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk"
		expectedChallenge := "E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM"

		challenge := ComputeCodeChallenge(verifier)
		if challenge != expectedChallenge {
			t.Fatalf("expected %s, got %s", expectedChallenge, challenge)
		}
	})
}

File: typescript/src/__tests__/pkce.test.ts
import { describe, expect, it } from 'vitest'
import {
  computeCodeChallenge,
  generateCodeVerifier,
  generateRandomString,
} from '../pkce.js'

describe('PKCE Utilities', () => {
  it('generates random string of specified length', () => {
    const s1 = generateRandomString(16)
    const s2 = generateRandomString(32)
    expect(s1).toHaveLength(16)
    expect(s2).toHaveLength(32)
    expect(s1).not.toBe(s2)
  })

  it('clamps code_verifier length between 43 and 128 characters', () => {
    const tooShort = generateCodeVerifier(10)
    const normal = generateCodeVerifier(64)
    const tooLong = generateCodeVerifier(200)

    expect(tooShort).toHaveLength(43)
    expect(normal).toHaveLength(64)
    expect(tooLong).toHaveLength(128)
  })

  it('computes S256 code_challenge correctly against RFC 7636 standard test vector', async () => {
    // Official test vector from RFC 7636 Appendix B
    const verifier = 'dBjftJeZ4CVP-mB92K27uhbUJU1p1r_wW1gFWFOEjXk'
    const expectedChallenge = 'E9Melhoa2OwvFrEMTJguCHaoeK1t8URWbuGJSstw-cM'

    const challenge = await computeCodeChallenge(verifier)
    expect(challenge).toBe(expectedChallenge)
  })
})
